README.md
# glmnet

![build passing](https://github.com/JoeLove100/glmnet/actions/workflows/standard_python.yml/badge.svg)

## Introduction
This is an implementation of the LocalGLMnet model suggested
by Ronald Richman and Mario W&uuml;thrich in their July 2021 paper
"*LocalGLMnet: interpretable deep learning for tabular data*". The idea behind 
this model is to create a framework for predictive analysis
on tabular data which combines the powerful modelling capabilities of 
feed-forward neuarl networks with the interpretability of GLMs.

## Using the package
The LocalGlmNet class essentially forms a wrapper around the basic keras
model, providing a convenient set of plotting functions to allow users to 
carry out feature importance and feature interaction analysis - the example
notebooks in the notebook directory provide more details on how this functionality can be used.
The .py files can be converted to a standard notebook using the 
[Jupytext package](https://github.com/mwouts/jupytext).

## Continuous variables

### Distribution of attention

The first plot that we can look at is the distribution of the attention beta weights against the actual
values of the different variables. This allows us to see how each feature impacts the final prediction, and
how this evolves as the value of the feature changes. 

We add horizontal lines to show confidence intervals for significance at the 95%, 99% and 99.9% level. These 
are based on the values observed for a totally random feature with no relation to the target variable whatsoever.

![plot](./resources/attention_weights.png)

### Feature interactions

By computing the derivative of each attention beta weight with respect to each of the variables, we can see
the way in which the different variables interact with one another.

![plot](./resources/interactions.png)

### Feature importance

Finally, we can take the mean attention beta weight to be the standalone "importance" of a given feature. Again,
we add a threshold based on the "importance" of a totally random variable.

![plot](./resources/feature_importance.png)

## Categorical variables

**TODO: add in charts for categorical variables**

## Original paper

A copy of the original paper can be found 
[here](https://arxiv.org/abs/2107.11059).


CHANGELOG.md
# Change log

| Version | Date | Description |
| ------- | ---- | ----------- |
| 1.0.0 | 2021-08-04 | Create initial package |

tests/test_local_glm_net.py
import unittest
import numpy as np
import numpy.random
from local_glm_net import LocalGlmNet


class TestLocalGlmNet(unittest.TestCase):

    @classmethod
    def _get_glm_model(cls):
        rng = np.random.default_rng(1234)
        model = LocalGlmNet(4, [2, 4], "regression", rng)
        return model

    def test_feature_to_plot_not_in_col_indices(self):
        # arrange
        model = self._get_glm_model()
        test_features_to_plot = ["name_1", "name_2", "name_3"]
        model.col_indices_ = {"name_1": 1, "name_3": 2}
        test_sample = 0.5

        # act/assert
        with self.assertRaises(ValueError):
            model.check_plot_arguments(test_features_to_plot, test_sample)

    def test_categorical_feature_not_in_col_indices(self):
        # arrange
        model = self._get_glm_model()
        test_features_to_plot = ["feature1", "feature2"]
        model.col_indices_ = {"feature1_val1": 1, "feature1_val2": 2}
        test_sample = 0.5

        # act/assert
        with self.assertRaises(ValueError):
            model.check_plot_arguments(test_features_to_plot, test_sample,
                                       is_categorical=True)

    def test_sample_negative_fails(self):
        # arrange
        model = self._get_glm_model()
        test_data = np.zeros(shape=(2, 2))
        test_feature_names = ["name_1", "name_2", "name_3"]
        model.col_indices_ = {"name_1": 1, "name_2": 2, "name_3": 3}
        test_sample = -0.1

        # act/assert
        with self.assertRaises(ValueError):
            model.check_plot_arguments(test_feature_names, test_sample)

    def test_sample_bigger_than_one_fails(self):
        # arrange
        model = self._get_glm_model()
        test_data = np.zeros(shape=(2, 2))
        test_feature_names = ["name_1", "name_2", "name_3"]
        model.col_indices_ = {"name_1": 1, "name_2": 2, "name_3": 3}
        test_sample = -0.1

        # act/assert
        with self.assertRaises(ValueError):
            model.check_plot_arguments(test_feature_names, test_sample)

    def test_get_sampled_data(self):
        # arrange
        test_data = np.array([[0.1, -0.2, 0.3],
                              [0.0, 0.1, -0.4],
                              [-0.3, 0.1, 0.7],
                              [0.4, 0.4, -0.2]])
        test_sample = 0.75
        model = self._get_glm_model()

        # act
        result = model.get_sampled_data(test_data, test_sample)

        # assert
        expected_result = np.array([[0.0, 0.1, -0.4, 0.1526191],
                                    [-0.3, 0.1, 0.7, 0.863744],
                                    [0.4, 0.4, -0.2, 2.913099]])
        np.testing.assert_array_almost_equal(expected_result, result)

    def test_get_confidence_intervals(self):
        # arrange
        model = self._get_glm_model()
        model.conf_ = 0.4

        # act
        result = model._get_confidence_intervals()

        # assert
        expected_result = {"95%": (-0.784, 0.784, "yellow"),
                           "99%": (-1.1228, 1.1228, "orange"),
                           "99.9%": (-1.3164, 1.3164, "red")}

        self.assertDictEqual(expected_result, result)


local_glm_net.py
import math
import numpy as np
import pandas as pd
import tensorflow as tf
from tensorflow import keras
from typing import List, Tuple, Dict, Optional
from matplotlib import pyplot as plt
from scipy.interpolate import UnivariateSpline

_VALID_ACTIVATIONS = ["relu", "tanh", "sigmoid", "linear"]
_VALID_MODEL_TYPES = {"regression": "linear",
                      "binary_classification": "sigmoid",
                      "poisson": "exponential"}


class LocalGlmNet:

    # region private methods
    def _create_keras_model(self) -> Tuple[keras.models.Model,
                                           keras.models.Model]:

        inputs = keras.Input(shape=self._shape, name="input")

        x = None
        for i, m in enumerate(self._layer_shapes):
            layer_name = f"hidden_{i}"
            if x is None:
                x = keras.layers.Dense(m, activation=self._layer_activation,
                                       name=layer_name)(inputs)
            else:
                x = keras.layers.Dense(m, activation=self._layer_activation,
                                       name=layer_name)(x)

        x = keras.layers.Dense(self._shape, activation="linear",
                               name="betas")(x)
        beta_model = keras.models.Model(inputs=inputs, outputs=x)
        linear = keras.layers.Dot(axes=[1, 1],
                                  name="linear")([x, inputs])
        final_activation = _VALID_MODEL_TYPES[self._model_type]
        outputs = keras.layers.Activation(final_activation,
                                          name="link")(linear)
        model = keras.models.Model(inputs=inputs, outputs=outputs,
                                   name="LocalGLMNet")

        # TODO: should allow parameters in the below to be
        #   customised by user rather than set automatically
        if self._model_type == "regression":
            model.compile(optimizer="adam", loss="mean_squared_error")
        elif self._model_type == "binary_classification":
            model.compile(optimizer="adam", loss="binary_crossentropy")
        elif self._model_type == "poisson":
            model.compile(optimizer="adam", loss="poisson")

        return model, beta_model

    def _get_confidence_intervals(self) -> Dict[str, Tuple[float, float, str]]:

        quantiles = [("95%", 1.960, "yellow"),
                     ("99%", 2.807, "orange"),
                     ("99.9%",  3.291, "red")]
        out = dict()
        for name, q, color in quantiles:
            out[name] = (-q * self.conf_, q * self.conf_, color)
        return out

    def get_sampled_data(self,
                         x_data: np.ndarray,
                         sample_size: float) -> np.ndarray:

        number_of_rows = int(x_data.shape[0] * sample_size)
        sample_rows = np.sort(self._rng.choice(x_data.shape[0],
                                               size=number_of_rows,
                                               replace=False))
        rand_col = self._rng.normal(size=(number_of_rows, 1))
        x_data_sample = x_data[sample_rows, :]
        x_data_sample = np.concatenate([x_data_sample, rand_col], axis=1)
        return x_data_sample

    def check_plot_arguments(self,
                             feature_names: List[str],
                             sample: float,
                             is_categorical: bool = False) -> None:

        # check sample proportion is valid
        if sample <= 0 or sample > 1:
            raise ValueError(f"Sample should be in interval (0, 1], "
                             f"but is {sample}")

        # check we actually have data for all features to plot
        if is_categorical:
            features_not_recognized = []
            for ft in feature_names:
                encoded_features = [col_name for col_name in self.col_indices_
                                    if col_name.startswith(ft)]
                if not encoded_features:
                    features_not_recognized.append(ft)
        else:
            features_not_recognized = [ft for ft in feature_names if
                                       ft not in self.col_indices_]

        if features_not_recognized:
            raise ValueError(f"The following features were not recognized: "
                             f"{', '.join(features_not_recognized)}")

    # endregion

    def __init__(self,
                 shape: int,
                 layer_shapes: List[int],
                 model_type: str,
                 random_generator: Optional[np.random.Generator] = None,
                 layer_activation: str = "relu") -> None:
        """
        wrapper class for the LocalGLMNet architecture with
        attached utility methods for feature selection/importance
        and visualising feature interactions

        :param shape: number of features in the data (ie the
        shape of input expected by the wrapped neural net)
        :param layer_shapes: the number of units in each of the
        hidden layers, the last of which should be of equal dimension
        to the model shape
        :param model_type: the form of the output of the model (ie
        classification, regression or poisson count)
        :param random_generator: numpy random number generator
        :param layer_activation: activation function for the hidden
        layers
        """

        if not layer_shapes:
            raise ValueError("Must provide shape for at least "
                             "one hidden layer")

        if layer_activation not in _VALID_ACTIVATIONS:
            raise ValueError(f"Activation function {layer_activation} "
                             f"is not supported: valid choices "
                             f"are {', '.join(_VALID_ACTIVATIONS)}")

        if model_type not in _VALID_MODEL_TYPES:
            raise ValueError(f"Model type {model_type} is not "
                             f"supported: valid choices "
                             f"are {', '.join(_VALID_MODEL_TYPES)}")

        if random_generator is None:
            random_generator = np.random.default_rng()

        # model parameters (adjust for random col)
        self._shape = shape + 1  # adjust for random col
        self._layer_shapes = layer_shapes
        self._model_type = model_type
        self._layer_activation = layer_activation
        self._rng = random_generator

        # only set after fitting
        self.prediction_model_ = None
        self.beta_model_ = None
        self.conf_ = None
        self.col_indices_ = {"Random": shape}

    def fit(self,
            x_train: pd.DataFrame,
            y_train: np.ndarray,
            val_split: float = 0.1,
            epochs: int = 100,
            use_early_stop: bool = True,
            **kwargs) -> None:
        """
        fit the wrapped neural network model using the
        standard Keras training loop

        :param x_train: numpy array of features to be used in fitting
        model weights
        :param y_train: numpy array of response values to be used in
        fitting model weights
        :param val_split: % of the training data that should be held back
        for validation purposes when training model weights
        :param epochs: number of training epochs
        :param use_early_stop: boolean flag for whether or not to add an
        early stopping callback
        """

        self.col_indices_.update({col: i for i, col in
                                  enumerate(x_train.columns)})
        x_train = x_train.values
        self.prediction_model_, self.beta_model_ = self._create_keras_model()
        random_col = self._rng.normal(size=(x_train.shape[0], 1))
        x_train = np.concatenate([x_train, random_col], axis=1)

        callbacks = []
        if use_early_stop:
            early_stop = keras.callbacks.EarlyStopping(monitor="val_loss",
                                                       patience=2)
            callbacks.append(early_stop)

        self.prediction_model_.fit(x=x_train, y=y_train, epochs=epochs,
                                   validation_split=val_split,
                                   callbacks=callbacks, **kwargs)

        betas = self.beta_model_(x_train)
        self.conf_ = np.std(betas[:, -1])

    def plot_betas_by_feature(self,
                              x_data: np.ndarray,
                              features_to_plot: Optional[List[str]] = None,
                              sample_size: float = 0.25,
                              cols: int = 3,
                              plot_random: bool = False,
                              plot_as_contributions: bool = False,
                              y_lim: Tuple[float, float] = (-1, 1)) \
            -> [plt.Figure, plt.Axes]:
        """
        utility function to plot the distribution of the local
        beta features given a set of feature values. Can optionally plot
        confidence intervals to help with feature selection,

        :param x_data: data for which we would like to plot our
        beta values
        :param features_to_plot: names for the features in our x_data
        that we would like to plot
        :param sample_size: % proportion of the feature values to sample
        to create our plots
        :param cols: number of columns in our plot
        :param plot_random: boolean flag for whether or not we would like
        to plot a figure for the additional random variable
        :param plot_as_contributions: boolean flag which, if true, plots
        contributions rather than outright beta values
        :param y_lim: y axis min and max values to scale plots properly
        """

        # get model betas
        if not features_to_plot and plot_random:
            features_to_plot = list(self.col_indices_)
        elif not features_to_plot:
            features_to_plot += [ft for ft in self.col_indices_ if
                                 ft != "Random"]
        elif plot_random:
            features_to_plot = features_to_plot + ["Random"]

        self.check_plot_arguments(features_to_plot, sample_size)
        x_data_sample = self.get_sampled_data(x_data, sample_size)
        betas = self.beta_model_(x_data_sample)
        col_indices = [self.col_indices_[ft] for ft in features_to_plot]
        betas = betas.numpy()[:, col_indices]
        x_data_sample = x_data_sample[:, col_indices]

        if plot_as_contributions:
            betas = x_data_sample * betas

        # set up our grid for plotting
        rows = betas.shape[1] // cols
        if betas.shape[1] % cols != 0:
            rows += 1
        fig, axs = plt.subplots(rows, cols)

        # set up our confidence intervals
        conf_intervals = self._get_confidence_intervals()

        # plot scatter chart for each variable
        for i, feature_name in enumerate(features_to_plot):

            r, c = i // cols, i % cols
            ax = axs[r, c]
            ax.scatter(x_data_sample[:, i], betas[:, i])
            if plot_as_contributions:
                ax.set_title(f"Contributions for {feature_name}")
            else:
                ax.set_title(f"Coefficients for {feature_name}")
            ax.set_ylim(*y_lim)

            if not plot_as_contributions:
                for name, definition in conf_intervals.items():
                    lower, upper, color = definition
                    ax.axhline(lower, color=color, linestyle="--")
                    ax.axhline(upper, color=color, linestyle="--")

        return fig, axs

    def plot_interactions(self,
                          x_data: np.ndarray,
                          features_to_plot: Optional[List[str]] = None,
                          sample_size: float = 1,
                          cols: int = 2) -> Tuple[plt.Figure, plt.Axes]:
        """
        for each feature we plot a graph showing how the gradients of
        the betas vary with the value of the given feature

        :param x_data: data for which we would like to plot our
        beta values
        :param features_to_plot: names for the features in our x_data
        that we would like to plot
        :param sample_size: % proportion of the feature values to sample
        to create our plots
        :param cols: number of columns in our plot
        """

        # calculate gradients based on sample
        if features_to_plot is None:
            features_to_plot = list(self.col_indices_)
        self.check_plot_arguments(features_to_plot, sample_size)
        x_data_sample = self.get_sampled_data(x_data, sample_size)
        input_tensor = tf.convert_to_tensor(x_data_sample)
        with tf.GradientTape() as tape:
            tape.watch(input_tensor)
            beta = self.beta_model_(input_tensor)
        grads = tape.batch_jacobian(beta, input_tensor)
        col_indices = [self.col_indices_[ft] for ft in features_to_plot]
        grads_np = grads.numpy()[np.ix_(range(grads.shape[0]),  col_indices,
                                        col_indices)]
        x_data_sample = x_data_sample[:, col_indices]

        # set up our axes for plotting
        rows = int(math.ceil(len(features_to_plot) / cols))
        fig, axs = plt.subplots(rows, cols)

        # outer loop: plot interactions for each feature
        for i, feature_name in enumerate(features_to_plot):

            # get axes for plotting
            row, col = i // cols, i % cols
            ax = axs[row, col]
            ax.set_ylim(-0.7, 0.7)  # TODO: change limits based on data
            ax.set_title(f"Interactions for feature {feature_name}",
                         fontsize=15)

            # select feature values and gradients, and reorder
            d_beta_0 = grads_np[:, i, :]
            x_0 = x_data_sample[:, i]
            order = np.argsort(x_0)
            x_0 = x_0[order]
            d_beta_0 = d_beta_0[order, :]
            x_vals = np.linspace(np.min(x_0), np.max(x_0), 30)

            # inner loop: plot each beta
            for j, other_feature_name in enumerate(features_to_plot):
                spline = UnivariateSpline(x_0, d_beta_0[:, j])
                ax.plot(x_vals, [spline(v) for v in x_vals],
                        label=other_feature_name)

            ax.legend(loc="lower right", ncol=2)

        return fig, axs

    def plot_feature_importance(self,
                                x_data: np.ndarray,
                                features_to_plot: Optional[List[str]] = None,
                                sample_size: float = 1) \
            -> Tuple[plt.Figure, plt.Axes]:
        """
        for each feature we plot a graph showing how the gradients of
        the betas vary with the value of the given feature

        :param x_data: data for which we would like to plot our
        beta values
        :param features_to_plot: names for the features in our x_data
        that we would like to plot
        :param sample_size: % proportion of the feature values to sample
        to create our plots
        """

        # get betas and calculate feature importance
        if features_to_plot is None:
            features_to_plot = list(self.col_indices_)
            features_to_plot.remove("Random")  # only plot random threshold

        self.check_plot_arguments(features_to_plot, sample_size)
        x_data_sample = self.get_sampled_data(x_data, sample_size)
        col_indices = [self.col_indices_[ft] for ft in features_to_plot] + [-1]
        betas = self.beta_model_(x_data_sample).numpy()[:, col_indices]
        avg_abs_betas = abs(betas[:-1]).mean(axis=0)
        importance, threshold = avg_abs_betas[:-1], avg_abs_betas[-1]

        # reorder data and feature names for plotting
        order = np.argsort(importance)
        importance = importance[order]
        features_to_plot = [features_to_plot[i] for i in order]

        # set up axes and make bar plot
        fig, ax = plt.subplots()
        positions = np.arange(len(features_to_plot))
        ax.barh(positions, importance)
        ax.set_yticks(positions)
        ax.set_yticklabels(features_to_plot)
        ax.axvline(threshold, color="red", linestyle="--")

        return fig, ax

    def plot_categorical_betas(self,
                               x_data: np.ndarray,
                               features_to_plot: Optional[List[str]] = None,
                               sample_size: float = 1,
                               cols: int = 2) \
            -> Tuple[plt.Figure, plt.Axes]:
        """
        plot box and whisker plots for each of the categorical
        variables. For this to work, categorical features must be
        one-hot encoded in the form {feature_name}_{value}

        :param x_data: data for which we would like to plot our
        beta values
        :param features_to_plot: names for the features in our x_data
        that we would like to plot
        :param sample_size: % proportion of the feature values to sample
        to create our plots
        :param cols: number of columns in resulting plot
        """

        # set features if not defined and check inputs
        if features_to_plot is None:
            features_to_plot = list(self.col_indices_)
            features_to_plot.remove("Random")  # only plot random threshold
        self.check_plot_arguments(features_to_plot, sample_size,
                                  is_categorical=True)

        # get data sample and use it to calculate betas
        x_data_sample = self.get_sampled_data(x_data, sample_size)
        betas = self.beta_model_(x_data_sample).numpy()

        # set up axes
        rows = int(math.ceil(len(features_to_plot) / cols))
        fig, axs = plt.subplots(rows, cols)

        # plot box and whisker for each feature
        for i, feature in enumerate(features_to_plot):

            r, c = i // cols, i % cols
            ax = axs[r, c]

            categories = [col_name for col_name in list(self.col_indices_)
                          if col_name.startswith(feature)]
            category_names = [c.split("_")[-1] for c in categories]
            category_indices = [self.col_indices_[c] for c in categories]
            betas_for_feature = betas[:, category_indices]
            ax.boxplot(betas_for_feature, labels=category_names)

        return fig, axs


notebooks/LocalGLMNet Example.py
# ---
# jupyter:
#   jupytext:
#     formats: ipynb,py:percent
#     text_representation:
#       extension: .py
#       format_name: percent
#       format_version: '1.3'
#       jupytext_version: 1.11.4
#   kernelspec:
#     display_name: glmnet
#     language: python
#     name: glmnet
# ---

# %% [markdown]
# # LocalGLMNet demonstration

# %% [markdown]
# The aim of this notebook is to provide a brief demonstration of the uses of the LocalGLMNet model. While such deep learning techniques are better suited to larger and more complex tabular datasets, we use the classic Abalone dataset here for simplicty.

# %%
# add parent directory to path

import sys
sys.path.append("../")

# %%
# do imports

import numpy as np
import pandas as pd
from typing import Tuple
from local_glm_net import LocalGlmNet
from matplotlib import pyplot as plt

# %% [markdown]
# ### Use dummy data

# %%
rng = np.random.default_rng(1994)

def get_dummy_target(row: pd.Series) -> float:
    
    x_1 = row.iloc[0]
    x_2 = row.iloc[1]
    x_3 = row.iloc[2]
    x_4 = row.iloc[3]
    x_5 = row.iloc[4]
    x_6 = row.iloc[5]
    
    val =  0.5 * x_1 - 0.25 * (x_2 ** 2) + 0.5 * abs(x_3) * np.sin(2 * x_3) + \
           0.5 * x_4 * x_5 + 0.125 * (x_5 ** 2) * x_6 + rng.normal()
    return val
    
def get_dummy_data(size: int) -> Tuple[pd.DataFrame, pd.Series]:
    
    corr = np.eye(8)
    corr[1, 7] = corr[7, 1] = 0.5
    features = rng.multivariate_normal(mean=np.zeros(8), cov=corr, size=size)
    features = pd.DataFrame(features, columns=[f"feature_{i + 1}" for i in range(8)])
    target = features.apply(get_dummy_target, axis=1)
    return features, target

dummy_features, dummy_target = get_dummy_data(int(1e5))

# %%
glm_model = LocalGlmNet(shape=8, layer_shapes=[20, 15, 10], 
                        model_type="regression", layer_activation="tanh")
glm_model.fit(dummy_features, dummy_target, epochs=200, val_split=0.2, verbose=False, batch_size=8)

# %%
fig, axs = glm_model.plot_betas_by_feature(dummy_features.values, sample_size=0.05, 
                                           plot_random=True, plot_as_contributions=False)
fig.set_size_inches(20, 8)
fig.tight_layout()

# %%
fig, axs = glm_model.plot_betas_by_feature(dummy_features.values, sample_size=0.05, 
                                           plot_random=True, plot_as_contributions=True)
fig.set_size_inches(20, 8)
fig.tight_layout()

# %%
fig, ax = glm_model.plot_interactions(dummy_features.values, sample_size=0.1)
fig.set_size_inches(15, 25)

# %%
fig, axs = glm_model.plot_feature_importance(dummy_features.values)
fig.set_size_inches(10, 10)

# %% [markdown]
# ### Read in and process Abalone data

# %% [markdown]
# We first need to read in the Abalone dataset from the TensorFlow datasets, and pre-process the features slighlty such that these are all normalised to have mean 0 and standard deviation 1.

# %%
# read in the raw data

raw_data = pd.read_csv('https://storage.googleapis.com/download.tensorflow.org/data/abalone_train.csv', 
                        names=["Length", "Diameter", "Height", "Whole weight", "Shucked weight", "Viscera weight", 
                               "Shell weight", "Age"])

# %%
# check for missing data

raw_data.isnull().sum()

# %%
# check summary of data

raw_data.describe().T

# %%
# remove height outliers

raw_data = raw_data[raw_data["Height"] < 0.5]

# %%
# split into features/target and normalise the former

data = raw_data.copy(deep=True)
features, target = data.drop("Age", axis=1), data["Age"]

for col in list(features):
    mean = features[col].mean()
    std = features[col].std()
    features[col] = (features[col] - mean) / std


# %%
features.describe().T

# %% [markdown]
# ### Create and fit our model

# %% [markdown]
# The next step is to create and fit our model. We go for a relatively simple model architecture with two hidden layers of size 16 and an output layer for our "attention" (beta) coefficients of size 7, in line with our features.  We then fit our model, ready to produce our GLM plots in the next sesion.  Note that the LocalGlmNet essenitally wraps a classic Keras neural network, and all training is handled by the standard Keras routines.

# %%
# define and fit model

glm_model = LocalGlmNet(shape=7, layer_shapes=[20, 15, 10, 40], 
                        model_type="regression", layer_activation="tanh")
glm_model.fit(features, target, epochs=200, verbose=False)

# %% [markdown]
# ### Create plots from model

# %% [markdown]
# Now we have a trained model we can use this to make a series of plots in order to better explore our data (and how our model uses it to make predictions). This is at the heart of "explainable" machine learning.  First of all, we can create a plot to show the distribution of the "attention" beta parameters for each feature at each point. Note the dotted lines, which show 95%, 99% and 99.9% central confidence intervals.

# %%
fig, axs = glm_model.plot_betas_by_feature(features.values, list(features), sample_size=1, plot_random=True, y_lim=(-12,12))
fig.set_size_inches(15, 20)
plt.show();

# %% [markdown]
# In our second chart, we can plot the "interactions" between different features, where we represent these as the gradient of each "attention" beta with respect to each of the underlying variables.

# %%
fig, axs = glm_model.plot_interactions(features.values, list(features), sample_size=1)
fig.set_size_inches(15, 20)
plt.show();

# %% [markdown]
# Finally, we can plot the feature importance, which we take to be the mean of the absolute values of the

# %%
fig, axs = glm_model.plot_feature_importance(features.values, list(features), sample_size=1)
fig.set_size_inches(8, 6)

# %%
